www/js/lib/filecache.js

```diff
@@ -0,0 +1,178 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Kiwix JS (file LICENSE).  If not, see <http://www.gnu.org/licenses/>
+ */
+'use strict';
+
+define(['q'], function (Q) {
+    /**
+     * Set maximum number of cache blocks of BLOCK_SIZE bytes each
+     * Maximum size of cache in bytes = MAX_CACHE_SIZE * BLOCK_SIZE
+     * @constant
+     * @type {Number}
+     */
+    const MAX_CACHE_SIZE = 4000;
+
+    /**
+     * The maximum blocksize to read or store via the block cache (bytes)
+     * @constant
+     * @type {Number}
+     */
+    const BLOCK_SIZE = 4096;
+
+    /**
+     * A Block Cache employing a Least Recently Used caching strategy
+     * @typedef {Object} BlockCache
+     * @property {Number} capacity The maximum number of entries in the cache 
+     * @property {Map} cache A map to store the cache keys and data
+     */
+
+    /**
+     * Creates a new cache with max size limit of MAX_CACHE_SIZE blocks
+     * LRUCache implemnentation with Map adapted from https://markmurray.co/blog/lru-cache/
+     */
+    function LRUCache() {
+        console.log('Creating cache of size ' + MAX_CACHE_SIZE + ' * ' + BLOCK_SIZE + ' bytes');
+        // Initialize persistent Cache properties
+        this.capacity = MAX_CACHE_SIZE;
+        this.cache = new Map();
+    }
+
+    /**
+     * Tries to retrieve an element by its id. If it is not present in the cache, returns undefined; if it is present,
+     * then the value is returned and the entry is moved to the bottom of the cache
+     * @param {String} key The block cache entry key (file.id + ':' + byte offset)
+     * @returns {Uint8Array | undefined} The requested cache data or undefined 
+     */
+    LRUCache.prototype.get = function (key) {
+        var entry = this.cache.get(key);
+        // If the key does not exist, return
+        if (!entry) return entry;
+        // Remove the key and re-insert it (this moves the key to the bottom of the Map: bottom = most recent)
+        this.cache.delete(key);
+        this.cache.set(key, entry);
+        // Return the cached data
+        return entry;
+    };
+
+    /**
+     * Stores a value in the cache by id and prunes the least recently used entry if the cache is larger than MAX_CACHE_SIZE
+     * @param {String} key The key under which to store the value (file.id + ':' + byte offset from start of ZIM archive)
+     * @param {Uint8Array} value The value to store in the cache 
+     */
+    LRUCache.prototype.store = function (key, value) {
+        var entry = this.cache.get(key);
+        // If the key already exists, delete it so that it will be added
+        // to the bottom of the Map (bottom = most recent)
+        if (entry) this.cache.delete(key);
+        else entry = value;
```

Comment:
I decided to keep the cached value for memory management purposes. Since the Map stores a *reference* to an object stored in memory (rather than storing a new object), then if we just overwrite the key with a new object (even with identical data), then we are not re-using memory optimally, and have to rely on the garbage collector to delete unlinked objects. I read somewhere that Maps hold on strongly to an object (as opposed to a [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap)).

So this explains why I decided to get the object reference from the map, and only store a new value if it returns `undefined`.

In practice, I think we will only very rarely look up in the ZIM a file slice that was already in the Cache: only where there is a race between two lookups requesting the same data before any one has completed and stored the data in the Cache. Once the Cache is substantially populated, this will be very rare.

But if you feel any of this is unsafe (or my reasoning flawed), I'll be happy to modify.

Reply:
No, it's fine, you are right.
But you should add these explanations in a comment, because this code looks weird at first sight